githubutil/githubutil.go

```diff
@@ -156,3 +156,92 @@ func DownloadFile(ctx context.Context, client *github.Client, owner, repo, branc
 
 	return []byte(content), nil
 }
+
+// FullyCreateFork creates a fork of the given repository under the PAT owner's account, waits for
+// the fork to be fully created, and returns its full information.
+func FullyCreateFork(ctx context.Context, client *github.Client, upstreamOwner, repo string) (*github.Repository, error) {
+	var fork *github.Repository
+
+	if err := Retry(func() error {
+		log.Printf("Creating fork of %v/%v...\n", upstreamOwner, repo)
+		var err error
+		fork, _, err = client.Repositories.CreateFork(ctx, upstreamOwner, repo, &github.RepositoryCreateForkOptions{
+			DefaultBranchOnly: true,
+		})
+		var acceptedError *github.AcceptedError
+		if errors.As(err, &acceptedError) {
+			return nil
+		}
+		return err
+	}); err != nil {
+		return nil, err
+	}
+
+	// Make sure the fork is done being created by getting the full info.
+	if err := Retry(func() error {
+		log.Printf("Checking for full fork information...\n")
+		var err error
+		fork, _, err = client.Repositories.Get(ctx, *fork.Owner.Login, *fork.Name)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+
+	return fork, nil
+}
+
+// PATScopes gets the scopes of the client's PAT by reading a simple API call's response headers.
+func PATScopes(ctx context.Context, client *github.Client) ([]string, error) {
+	var scopes []string
+
+	if err := Retry(func() error {
+		log.Printf("Fetching PAT scopes...\n")
+		_, resp, err := client.Meta.Get(ctx)
+		if err != nil {
+			return err
+		}
+		for _, s := range strings.Split(resp.Header.Get("X-OAuth-Scopes"), ",") {
+			scopes = append(scopes, strings.TrimSpace(s))
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return scopes, nil
+}
+
+// FetchRepositoryOrNil fetches a repository from GitHub, or returns nil if it doesn't exist.
+// Returning nil simplifies error handling for the caller if it's expected that the repository may
```

Comment:
One the other hand, it complicates the error handling if the repo is expected to exists. Callers will have to check for two conditions: `err != nil && repo != nil`, and they better don't forget the latter, else they will have a nil pointer exception down the line.

The idiomatic and less error-prone way to do this is defining an error sentinel value that is returned only when the repo doesn't exist, see for example [os.ErrNotExist](https://pkg.go.dev/os#ErrNotExist) and [sql.ErrNoRows](https://pkg.go.dev/database/sql#ErrNoRows). This way callers can rely just on the error to implement error-handling logic.

Reply:
I expected that callers expecting the repo to exist will simply use `client.Repositories.Get`, which does err out. 😄 But yeah, on second thought, there's no reason not to code this as a more general-purpose wrapper. (I wrote a very similar comment for the `DownloadFile` implementation originally, although it was already being used as a general-purpose wrapper.)